### PR TITLE
Add go licenses to licenses.txt

### PR DESCRIPTION
**Pull request by silverwind**
`make go-licenses` will generate `assets/go-licenses.json` which is then included in the webpack build. 

This step depends on both go and node being present, so unfortunately, I could not automate the generation by hooking it up to `tidy` as that target is triggered on CI where we do not have a docker image with both go an node.

It should be ran from time to time, ideally after each go mod update.

### CODE REVIEW

**Comment by 6543** [comment]
`coverage-bodged.out` should not be within this pull

**Comment by silverwind** [comment]
Should be fixed, branch got messed up through rebases.

File: build/generate-go-licenses.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+});
+
+function exit(err) {
+  if (err) console.error(err);
+  process.exit(err ? 1 : 0);
+}
+
+async function main() {
+  const line = '-'.repeat(80);
+  const str = glob(['**/*']).filter((path) => {
```
Comment (silverwind) [inline build/generate-go-licenses.js]:
```suggestion
  const str = glob('**/*').filter((path) => {
```

**Review by lunny — APPROVED** [review]

**Review by delvh — APPROVED** [review]

**Comment by silverwind** [comment]
Small rewrite incoming, which will sort all modules at once.

**Comment by silverwind** [comment]
Data is now passed through a JSON file to webpack, which enabled me to sort all modules at once, e.g. showing now preference for js vs. go modules like earlier.

**Comment by silverwind** [comment]
One future addition might be to parse the go module version from `go.sum`, but I'm not feeling like that right now 😉.

**Review by delvh — APPROVED** [review]

**Comment by 6543** [comment]
@silverwind add `go-licenses` target to `make tidy` and `make vendor` ?

**Comment by silverwind** [comment]
Possible, but it would require node to be installed for `checks-backend` ci target.

**Comment by silverwind** [comment]
did that now, so `make tidy` now triggers the target if necessary. I didn't feel like adding it to `make vendor` because that is invoked in release pipelines and it seems just unnecessary there.

**Review by 6543 — APPROVED** [review]

**Comment by zeripath** [comment]
This is still running on make vendor and it's running somewhat intermittently occasionally resulting in an empty file.

I would say that this should really be in make generate as it is a generation step not tidy.

**Comment by silverwind** [comment]
> This is still running on make vendor and it's running somewhat intermittently occasionally resulting in an empty file.
> 
> I would say that this should really be in make generate as it is a generation step not tidy.

It has to run after `tidy` thought, otherwise it may produce outdated results.

**Comment by zeripath** [comment]
> It has to run after `tidy` thought, otherwise it may produce outdated results. Thought if it runs after `tidy-check`, that will also work because that target ensure tidy is consistent.

You can't really build without `make vendor` being up-to-date so that's probably not the issue.

In fact we should be able to make make `make vendor` when `make backend` if go.mod is newer than go.sum, and then if we make `make go-licenses` dependent on vendor and go.mod we should be able to make make `make go-licenses` when we `make frontend` 

The only issue is how drone will behave.

**Comment by silverwind** [comment]
As I see it `make vendor` should be a complete standalone task. It just downloads what is in `go.sum` to the `vendor`directory, used for release source tarballs. Build will work with or without this directory as it sources dependencies from `GOROOT`.

In any case, I think I will at least rewrite `generate-go-licenses.js` to go, to eliminate the node dependency of the step.

**Comment by silverwind** [comment]
Go rewrite of script in https://github.com/go-gitea/gitea/pull/21078.

**Pull request merged**